GitUI/UserControls/RevisionGrid/RevisionGridMenuCommands.cs

```diff
@@ -421,5 +433,35 @@ public void GotoCommitExecute()
                 }
             }
         }
+
+        public void SwitchBetweenArtificialCommits()
+        {
+            _revisionGrid?.GoToRef(GetIdToSelect()?.ToString(), showNoRevisionMsg: false);
+            OnSwitchBetweenArtificialCommits?.Invoke(this, EventArgs.Empty);
+            return;
+
+            ObjectId GetIdToSelect()
+            {
+                var currentCheckout = _revisionGrid?.CurrentCheckout;
+
+                // try the up to 3 possibilities
+                ObjectId idToSelect = _revisionGrid?.LatestSelectedRevision?.ObjectId;
+                for (int i = 0; i < 3; ++i)
+                {
+                    idToSelect = GetNextIdToSelect(idToSelect);
+                    if (idToSelect != null && (!idToSelect.IsArtificial || (_revisionGrid?.GetChangeCount(idToSelect)?.Any ?? false)))
+                    {
+                        return idToSelect;
+                    }
+                }
+
+                return currentCheckout;
+
+                ObjectId GetNextIdToSelect(ObjectId id)
+                    => id == ObjectId.WorkTreeId ? ObjectId.IndexId
+                     : id == ObjectId.IndexId ? currentCheckout
+                     : ObjectId.WorkTreeId;
+            }
```

Comment:
This method belong to `_revisionGrid`, as it incapsulates behaviours that occur in the revision grid.

Reply:
Have you noticed the file and class name: "RevisionGrid/RevisionGridMenuCommands.cs"?
This function is an extension to `RevisionGrid`, i.e. no need to bloat the control class with this logic.
Why should this function be handled differently than the function `RevisionGridMenuCommands.GotoCommitExecute` right above?

Reply:
I have. But it doesn't explain why we need to add another layer to perform an action in `RevisionGrid`.
The commits are owned by `RevisionGrid`, `RevisionGridMenuCommands` managed the context menu that belong to the grid. Now `RevisionGridMenuCommands` performs switching on commits in the grid, defines an event that is fired when an action is performed in the grid, instead of the grid. Then the grid has to send a reaction to a shortcut to `RevisionGridMenuCommands` instead of reacting itself.... It just feels confusing.

```diff
// FormBrowse
- RevisionGrid.MenuCommands.ToggledBetweenArtificialAndHeadCommits += (s, e) => FocusRevisionDiffFileStatusList();
+ RevisionGrid.ToggledBetweenArtificialAndHeadCommits += (s, e) => FocusRevisionDiffFileStatusList();


// RevisionGrid.cs
+        /// <summary>
+        ///  Occurs whenever a user toggles between the artificial and the HEAD commits
+        ///  via the navigation menu item or the shortcut command.
+        /// </summary>
+        public event EventHandler ToggledBetweenArtificialAndHeadCommits;

- public bool HasChanges
+ private bool HasChanges

-  case Command.ToggleBetweenArtificialAndHeadCommits: MenuCommands.ToggleBetweenArtificialAndHeadCommits(); break;
+  case Command.ToggleBetweenArtificialAndHeadCommits: ToggleBetweenArtificialAndHeadCommits(); break;


+        public void ToggleBetweenArtificialAndHeadCommits()
+        {
+            ....
+        }
````